### ISSUE TITLE
Support metadataCharset on Project.zipTree()

### ISSUE THREAD

**Issue by oehme** [gradle/gradle#1363]
Our Zip and Jar tasks support setting the metadata charset when creating archives. But when Gradle consumes a Zip file using `Project.zipTree()`, it always uses the platform encoding.

### Expected Behavior
`Project.zipTree()` can be configured with the right metadata encoding. 

Alternatives:

- always use UTF-8 as the default (closer to the ZIP spec which doesn't really allow arbitrary encodings - though everybody ignores that)
- use UTF-8 when the file extension is `.jar`, since the JVM expects jars to be UTF-8 encoded. This would mirror the default on the `Jar` task

### Current Behavior
`Project.zipTree()` always uses the platform encoding, which breaks when a ZIP has been UTF-8 encoded and someone tries to expand it on Windows for instance.

### Context
https://github.com/javaslang/javaslang/issues/1856

**Comment by melix** [comment]
That's not what the Javadoc says for `ZipFile`:

> Opens a ZIP file for reading given the specified File object.
> The UTF-8 charset is used to decode the entry names and comments.

The version which accepts an encoding needs to be used when you read a Zip file which entries are *not* using UTF-8.

**Comment by oehme** [comment]
You're right, but our `ZipFileTree` uses `org.apache.tools.zip.ZipFile` instead, which uses the platform encoding by default.

**Comment by melix** [comment]
Oh I missed the fact it wasn't the same `ZipFile`. Thanks for the clarification.

**Comment by eskatos** [comment]
Here is the design doc we used to fix archive metadata charset handling in 3.x: https://github.com/gradle/gradle/blob/master/design-docs/handling-encodings.md

The approach we took was to:
- use UTF-8 by default for JAR, WAR and EAR as using another charset is incorrect per the spec ;
- keep using the default platform charset for ZIP metadata for backward compatibility reasons, using the *Language encoding flag* (or EFS).

Missing `Project.zipTree` was an overlook.
We should at least make it consistent with the archive tasks.

We could also consider using UTF-8 by default for ZIPs in 4.0 as a potentially breaking change.

Note that this also applies to `Project.tarTree`.
